vm/vmm/vmm.go

```diff
@@ -0,0 +1,422 @@
+	vmDst := filepath.Join("/root", filepath.Base(hostSrc))
+	args := append(vmimpl.SCPArgs(inst.debug, inst.sshkey, inst.sshport), hostSrc, inst.sshuser+"@"+inst.sshhost+":"+vmDst)
+	cmd := osutil.Command("scp", args...)
+	if inst.debug {
+		log.Logf(0, "running command: scp %#v", args)
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stdout
+	}
+	if err := cmd.Start(); err != nil {
+		return "", err
+	}
+	done := make(chan bool)
+	go func() {
+		select {
+		case <-time.After(3 * time.Minute):
+			cmd.Process.Kill()
+		case <-done:
+		}
+	}()
+	err := cmd.Wait()
+	close(done)
+	if err != nil {
+		return "", err
+	}
+	return vmDst, nil
+}
+
+func (inst *instance) Run(timeout time.Duration, stop <-chan bool, command string) (
+	<-chan []byte, <-chan error, error) {
+	rpipe, wpipe, err := osutil.LongPipe()
+	if err != nil {
+		return nil, nil, err
+	}
+	inst.merger.Add("ssh", rpipe)
+
+	args := append(vmimpl.SSHArgs(inst.debug, inst.sshkey, inst.sshport), inst.sshuser+"@"+inst.sshhost, command)
+	if inst.debug {
+		log.Logf(0, "running command: ssh %#v", args)
+	}
+	cmd := osutil.Command("ssh", args...)
+	cmd.Stdout = wpipe
+	cmd.Stderr = wpipe
+	if err := cmd.Start(); err != nil {
+		wpipe.Close()
+		return nil, nil, err
+	}
+	wpipe.Close()
+	errc := make(chan error, 1)
+	signal := func(err error) {
+		select {
+		case errc <- err:
+		default:
+		}
+	}
+
+	go func() {
+		select {
+		case <-time.After(timeout):
+			signal(vmimpl.ErrTimeout)
+		case <-stop:
+			signal(vmimpl.ErrTimeout)
+		case err := <-inst.merger.Err:
+			cmd.Process.Kill()
+			if cmdErr := cmd.Wait(); cmdErr == nil {
+				// If the command exited successfully, we got EOF error from merger.
+				// But in this case no error has happened and the EOF is expected.
+				err = nil
+			}
+
+			signal(err)
+			return
+		}
+		cmd.Process.Kill()
+		cmd.Wait()
+	}()
+	return inst.merger.Output, errc, nil
+}
+
+func (inst *instance) Diagnose() bool {
+	didWrite := false
+	commands := []string{"", "trace", "show registers"}
+	for _, c := range commands {
+		if inst.wpipe != nil {
+			didWrite = true
+			inst.wpipe.Write([]byte(c + "\n"))
+			time.Sleep(100 * time.Millisecond)
+		}
+	}
+	return didWrite
+}
+
+func (inst *instance) console() {
+	run := func() error {
+		outr, outw, err := osutil.LongPipe()
+		if err != nil {
+			return fmt.Errorf("failed to create output pipe: %v", err)
+		}
+		inr, inw, err := osutil.LongPipe()
+		if err != nil {
+			return fmt.Errorf("failed to create input pipe: %v", err)
+		}
+		inst.wpipe = inw
+
+		cmd := osutil.Command("vmctl", "console", inst.vmIdent())
+		cmd.Stdin = inr
+		cmd.Stdout = outw
+		cmd.Stderr = outw
+		if err := cmd.Start(); err != nil {
+			return err
+		}
+		outw.Close()
+		inr.Close()
+		inst.merger.Add("console", outr)
+
+		if _, err := cmd.Process.Wait(); err != nil {
+			return err
+		}
+		inst.wpipe = nil
+		inw.Close()
+		outr.Close()
+		return nil
+	}
+	go func() {
+		for inst.alive() {
```

Comment:
Why do we start this in the loop? Why can't we start it just once?

Reply:
I've seen the console process die after executing some programs, not sure why yet. However the VM is still in a working state therefore the retry mechanism.